src/Random.mo

```diff
@@ -8,63 +18,206 @@ module {
 
   public let blob : shared () -> async Blob = rawRand;
 
-  /// Opinionated choice of PRNG from a given seed.
-  public func new(seed : Blob) : Random {
-    todo()
+  /// Creates a fast pseudo-random number generator using the SFC64 algorithm.
+  /// This provides statistical randomness suitable for simulations and testing,
+  /// but should not be used for cryptographic purposes.
+  /// The seed blob's first 8 bytes are used to initialize the PRNG.
+  public func fast(seed : Nat64) : Random {
+    let prng = PRNG.sfc64a();
+    prng.init(seed);
+    Random(
+      func() {
+        // Generate 8 bytes directly from a single 64-bit number
+        let n = prng.next();
+        let bytes = VarArray.repeat<Nat8>(0, 8);
+        bytes[0] := Nat8.fromNat(Nat64.toNat(n & 0xFF));
+        bytes[1] := Nat8.fromNat(Nat64.toNat((n >> 8) & 0xFF));
+        bytes[2] := Nat8.fromNat(Nat64.toNat((n >> 16) & 0xFF));
+        bytes[3] := Nat8.fromNat(Nat64.toNat((n >> 24) & 0xFF));
+        bytes[4] := Nat8.fromNat(Nat64.toNat((n >> 32) & 0xFF));
+        bytes[5] := Nat8.fromNat(Nat64.toNat((n >> 40) & 0xFF));
+        bytes[6] := Nat8.fromNat(Nat64.toNat((n >> 48) & 0xFF));
+        bytes[7] := Nat8.fromNat(Nat64.toNat((n >> 56) & 0xFF));
+        Blob.fromArray(Array.fromVarArray(bytes))
+      }
+    )
   };
 
-  /// Uses entropy from the management canister with automatic resupply.
-  public func newAsync() : AsyncRandom {
-    todo()
+  /// Creates a random number generator suitable for cryptography
+  /// using entropy from the ICP management canister with automatic resupply.
+  public func crypto() : AsyncRandom {
+    AsyncRandom(func() : async* Blob { await rawRand() })
   };
 
   public class Random(generator : () -> Blob) {
+    var iter : Iter.Iter<Nat8> = Iter.empty();
+    let bitIter : Iter.Iter<Bool> = object {
+      var mask = 0x00 : Nat8;
+      var byte = 0x00 : Nat8;
+      public func next() : ?Bool {
+        if (0 : Nat8 == mask) {
+          switch (iter.next()) {
+            case null { null };
+            case (?w) {
+              byte := w;
+              mask := 0x40;
+              ?(0 : Nat8 != byte & (0x80 : Nat8))
+            }
+          }
+        } else {
+          let m = mask;
+          mask >>= (1 : Nat8);
+          ?(0 : Nat8 != byte & m)
+        }
+      }
+    };
 
     /// Random choice between `true` and `false`.
     public func bool() : Bool {
-      todo()
+      switch (bitIter.next()) {
+        case (?bit) { bit };
+        case null {
+          iter := generator().vals();
+          switch (bitIter.next()) {
+            case (?bit) { bit };
+            case null {
+              Runtime.trap("Random.bool(): generator produced empty Blob")
+            }
+          }
+        }
+      }
     };
 
     /// Random `Nat8` value in the range [0, 256).
-    public func byte() : Nat8 {
-      todo()
+    public func nat8() : Nat8 {
+      switch (iter.next()) {
```

Comment:
I guess this mean we will discard the remaining bits in bitIter, right?
Maybe just add a TODO to use them in future? Or extend the bitIter api to give you the remaining bits?

Reply:
@ggreif, do you know if this was intentional? I was thinking about this as well (resetting the bit mask/byte on refill), but perhaps there was a performance reason for this.

Reply:
I think it was intentional. I was contemplating having a single-bit mask being shifted in the iterator and skipping to the next byte when empty (maybe I have even implemented this somewhere?). OTOH, it would reduce the resupplying by a factor of 8...